pennylane/beta/tapes/tape.py

```diff
@@ -0,0 +1,835 @@
+        stop_at (Callable): A function which accepts a queue object,
+            and returns ``True`` if this object should *not* be expanded.
+            If not provided, all objects that support expansion will be expanded.
+        expand_measurements (bool): If ``True``, measurements will be expanded
+            to basis rotations and computational basis measurements.
+
+    **Example**
+
+    Consider the following nested tape:
+
+    .. code-block:: python
+
+        with QuantumTape() as tape:
+            qml.BasisState(np.array([1, 1]), wires=[0, 'a'])
+
+            with QuantumTape() as tape2:
+                qml.Rot(0.543, 0.1, 0.4, wires=0)
+
+            qml.CNOT(wires=[0, 'a'])
+            qml.RY(0.2, wires='a')
+            probs(wires=0), probs(wires='a')
+
+    The nested structure is preserved:
+
+    >>> tape.operations
+    [BasisState(array([1, 1]), wires=[0, 'a']),
+     <QuantumTape: wires=[0], params=3>,
+     CNOT(wires=[0, 'a']),
+     RY(0.2, wires=['a'])]
+
+    Calling ``expand_tape`` will return a tape with all nested tapes
+    expanded, resulting in a single tape of quantum operations:
+
+    >>> new_tape = expand_tape(tape)
+    >>> new_tape.operations
+    [PauliX(wires=[0]),
+     PauliX(wires=['a']),
+     Rot(0.543, 0.1, 0.4, wires=[0]),
+     CNOT(wires=[0, 'a']),
+     RY(0.2, wires=['a'])]
+    """
+    if depth == 0:
+        return tape
+
+    if stop_at is None:
+        # by default expand all objects
+        stop_at = lambda obj: False
+
+    new_tape = tape.__class__()
+
+    for queue in ("_prep", "_ops", "_measurements"):
+        for obj in getattr(tape, queue):
+
+            stop = stop_at(obj)
+
+            if not expand_measurements:
+                # Measurements should not be expanded; treat measurements
+                # as a stopping condition
+                stop = stop or isinstance(obj, MeasurementProcess)
+
+            if stop:
+                # do not expand out the object; append it to the
+                # new tape, and continue to the next object in the queue
+                getattr(new_tape, queue).append(obj)
+                continue
+
+            if isinstance(obj, (qml.operation.Operation, MeasurementProcess)):
+                # Object is an operation; query it for its expansion
+                try:
+                    obj = obj.expand()
+                except NotImplementedError:
+                    # Object does not define an expansion; treat this as
+                    # a stopping condition.
+                    getattr(new_tape, queue).append(obj)
+                    continue
+
+            # recursively expand out the newly created tape
+            expanded_tape = expand_tape(obj, stop_at=stop_at, depth=depth - 1)
+
+            new_tape._prep += expanded_tape._prep
+            new_tape._ops += expanded_tape._ops
+            new_tape._measurements += expanded_tape._measurements
+
+    return new_tape
```

Comment:
Nice. I remember when learning F# I thought quite often that recursion is the single most underutilised programming paradigm in python!